dvc/repo/ls.py

```diff
@@ -0,0 +1,109 @@
+
+
+@staticmethod
+def ls(url, targets=[], rev=None, recursive=None):
+    if os.path.exists(url):
+        return _ls_local_repo(url, targets, recursive)
+
+    return _ls_git_repo(url, targets, rev, recursive)
+
+
+def _ls_files_local_repo(path, targets=[], recursive=None):
+    from dvc.repo import Repo
+
+    files = []
+    try:
+        repo = Repo(path)
+        if not len(targets):
+            targets.append(".")
+
+        def prettify_file(prefix, filename, is_root):
+            if not is_root:
+                filename = os.path.join(prefix, filename)
+            return filename
+
+        def prettify_dir(prefix, filename, is_root):
+            filepath = prettify_file(prefix, filename, is_root)
+            return filepath + "/"
+
+        for target in targets:
+            target_path = os.path.join(path, target)
+            if not os.path.exists(target_path):
+                continue
+
+            for dirpath, dirnames, filenames in repo.tree.walk(target_path):
+                prefix = os.path.relpath(dirpath, path)
+                is_root = os.path.abspath(dirpath) == os.path.abspath(path)
+                files.extend(
+                    map(lambda f: prettify_dir(prefix, f, is_root), dirnames)
+                )
+                files.extend(
+                    map(lambda f: prettify_file(prefix, f, is_root), filenames)
+                )
+                if not recursive:
+                    break
+
+    except NotDvcRepoError:
+        pass
+    finally:
+        repo.close()
+    return files
+
+
+def _ls_outs_local_repo(path, targets=[], recursive=None):
+    from dvc.repo import Repo
+    from dvc.path_info import PathInfo
+
+    repo = Repo(path)
+    if not len(targets):
+        targets.append(path)
+    outs = set()
+    for stage in repo.stages:
+        outs = outs.union(stage.outs)
+
+    def in_target(out, target_path_info, is_dir):
+        if out.scheme == "local" and out.path_info == target_path_info:
+            return True
+
+        if is_dir:
+            if recursive and out.path_info.isin(target_path_info):
+                return True
+            if out.path_info.isin(target_path_info, 1):
+                return True
+
+        return False
+
+    matched_outs = []
+    for path in targets:
+        abs_path = os.path.abspath(path)
+        path_info = PathInfo(abs_path)
+        is_dir = os.path.isdir(abs_path)
+        matched = set()
+        for out in outs:
+            if in_target(out, path_info, is_dir):
+                matched.add(out)
+                matched_outs.append(out)
+        outs.difference_update(matched)
+
+    return matched_outs
+
+
+def _ls_local_repo(path, targets=[], recursive=None):
+    path = os.path.abspath(path)
+    files = _ls_files_local_repo(path, targets, recursive)
+    outs = _ls_outs_local_repo(path, targets, recursive)
+    result = files + list(map(str, outs))
+    result = list(set(result))
+    result.sort()
+    return result
```

Comment:
```python
return sorted(set(result))
```
Also, why set? Do we expect them to be duplicate?

Reply:
Let me check, I remember that faced with duplicated rows, but maybe it was an issue in the logic if outs cannot be duplicated

Reply:
I've got how it can happen - the output can  be fetched/generated and exists on the fs. Hence it will be returned from `_ls_files` and from `_ls_outs` also which leads to duplicates in the output